fdmi/fol_fdmi_src.c

```diff
@@ -104,6 +104,96 @@
+ * gracefully (graceful shutdown/startup). In either case there may be FDMI FOL
+ * records that don't have consumption confirmation from FDMI plugin. Current
+ * implementation takes BE tx reference until there is such confirmation from
+ * FDMI plugin. BE tx reference taken means in this case that BE tx wouldn't be
+ * discarded from BE log until the reference is put. BE recovery recovers all
+ * transactions that were not discarded, which is very useful for FDMI FOL
+ * record resend case: if all FOL records for such recovered transactions are
+ * resent as FDMI FOL records then there will be no FDMI FOL record missing on
+ * FDMI plugin side regardless whether FDMI source restarts or not, how and how
+ * many times it restarts.
+ *
+ * It leads to an obvious solution: just send all FOL records as FDMI FOL
+ * records during BE recovery.
+ *
+ * There are several ways this task could be done.
+ *
+ * @subsection Original FOM for each FOL record
+ *
+ * For each FOL record an orignal fom is created. A flag about being created
+ * during BE recovery is passed to each such fom, so fom wouldn't do it's usual
+ * action. It will transition through generic phases that involve BE tx. After
+ * BE tx goes to M0_BTS_LOGGED phases a generic phase executes
+ * m0_fom_fdmi_record_post(), which would send FDMI record to FDMI plugin as
+ * usual.
+ *
+ * @subsection Special FOM for each FOL record
+ *
+ * A special FOM would be created for each recovered BE tx. The purpose of the
+ * fom would be post FDMI record with the FOL record for this BE tx and then
+ * wait until consumption of the FDMI record is acklowledged.
+ *
+ * @subsection Post FDMI records for every BE tx
+ *
+ * FDMI FOL record for every recovered BE tx is posted as usual. FDMI source
+ * dock fom would make a queue of all the records and it would send them and
+ * wait for consumption acknowledgement as usual.
+ *
+ * @subsection A special BE recovery FOM phase
+ *
+ * A special phase that indicates that this fom is created during BE recovery is
+ * added to each fom and is initial phase in case if the fom is created during
+ * BE recovery. This allows each fom to handle BE recovery as it sees fit.
+ * Default generic phase sequence would post FDMI FOL record in the same way
+ * it's done currently for normal FOM phase sequence.
+ *
+ * @subsection Implementation details
+ *
+ * - Motr process should be able to send FDMI fops during BE recovery. BE
+ *   recovery happens before DTM recovery, so at this stage only HA and Motr
+ *   configuration should be brought up (along with their dependencies). It
+ *   means that ioservice and DIX would be down at that time;
+ * - FOL record shouldn't be discarded until it's consumed by FDMI plugin.
+ *   Currently FOL record is in BE tx payload, so this requriement means that BE
+ *   tx shouldn't be discarded from BE log before FDMI FOL record is consumed by
+ *   FDMI plugin. It has several side effects:
```

Comment:
What shall we do?
Several possible implementations, with pros and cons.
It is not decided which is the best solution, right?
So, we still need more investigation, and experiments?

Reply:
Yes, more investigation may be needed. I haven't seen new Motr design documents about interacting with the rest of the system (do we have any?), so I don't know which way of doing this would be fit LR R2 design in the best way.

There is nothing to experiment with at this point.

Reply:
OK